File: clinic_backend/src/services/invoiceService.ts
import prisma from '../config/prisma.js';
import { ApiError } from '../utils/apiError.js';
import { CreateInvoiceInput, payInvoiceInput } from '../validation/invoiceSchema.js';
import { getVisitByIdService } from './visitService.js';
import { InvoiceStatus, PaymentMethod } from '@prisma/client';
import { io } from '../app.js';
import { snap } from '../config/midtrans.js';

const invoiceSelectPayload = {
  // 1. Field Utama Invoice (Faktur Kasir)
  id: true,
  invoiceNo: true,
  totalConsultationFee: true,
  totalMedicineFee: true,
  totalAmount: true,
  status: true,
  paymentMethod: true,
  paidAt: true,
  createdAt: true,

  // 2. Relasi ke Visit (Data Pasien, Dokter, Diagnosa, & Rincian Obat)
  visit: {
    select: {
      id: true,
      visitDate: true,
      patient: {
        select: {
          id: true,
          noRm: true,
          name: true,
          phone: true,
        },
      },
      doctor: {
        select: {
          id: true,
          name: true,
          spesialis: true,
        },
      },
      consultation: {
        select: {
          id: true,
          complaint: true,
          diagnosis: true,
          notes: true,
          consultationMedicines: {
            select: {
              id: true,
              qty: true,
              price: true,
              subTotal: true,
              instructions: true,
              medicine: {
                select: {
                  id: true,
                  code: true,
                  name: true,
                  unit: true,
                },
              },
            },
          },
        },
      },
    },
  },
} as const;

export const getAllInvoiceService = async () => {
  const invoices = await prisma.invoice.findMany({
    select: {
      id: true,
      invoiceNo: true,
      totalConsultationFee: true,
      totalMedicineFee: true,
      totalAmount: true,
      status: true,
      paymentMethod: true,
      paidAt: true,
      visit: {
        select: {
          patient: { select: { name: true } },
          doctor: { select: { name: true } },
        },
      },
    },
  });

  if (invoices.length === 0) {
    throw new ApiError(404, 'Invoice nya kosong');
  }

  return invoices;
};

export const createInvoiceService = async (input: CreateInvoiceInput) => {
  await getVisitByIdService(input.visitId);

  // cek dlu apakah pasien sudah punya dokter
  const consultation = await prisma.consultation.findUnique({
    where: { visitId: input.visitId },
    include: { consultationMedicines: true },
  });

  if (!consultation) {
    throw new ApiError(404, 'Pasien ini belum di periksa oleh Dokter');
  }

  //cek apakah invoice sudah pernah di buat sebelumnya
  const existingInvoice = await prisma.invoice.findUnique({
    where: { visitId: input.visitId },
  });

  if (existingInvoice) {
    throw new ApiError(400, 'Invoice ini sudah pernah dibuat sebelumnya');
  }

  const totalConsultationFee = Number(consultation.consultationFee);
  const totalMedicineFee = consultation.consultationMedicines.reduce((sum, item) => sum + Number(item.subTotal), 0);
  const totalAmount = totalConsultationFee + totalMedicineFee;

  const countAll = await prisma.invoice.count();
  const invoiceNo = `INV-${String(countAll + 1).padStart(4, '0')}`;

  const createInvoice = await prisma.invoice.create({
    data: {
      visitId: input.visitId,
      invoiceNo: invoiceNo,
      totalConsultationFee: totalConsultationFee,
      totalMedicineFee: totalMedicineFee,
      totalAmount: totalAmount,
      paymentMethod: input.paymentMethod ?? PaymentMethod.CASH,
    },
  });
  io.emit('QUEUE_UPDATED', { type: 'INVOICE_CREATED', invoiceId: createInvoice.id });

  return createInvoice;
};

export const getInvoiceByIdService = async (id: number) => {
  const getInvoice = await prisma.invoice.findUnique({
    where: { id },
    select: invoiceSelectPayload,
  });

  if (!getInvoice) {
    throw new ApiError(404, 'Data Invoice tidak di temukan');
  }

  return getInvoice;
};

export const payInvoiceService = async (id: number, input?: payInvoiceInput) => {
  const invoice = await getInvoiceByIdService(id);

  if (invoice.status === InvoiceStatus.PAID) {
    throw new ApiError(400, 'Invoice ini sudah lunas sebelumnya');
  }

  const payInvoice = await prisma.invoice.update({
    where: { id },
    data: {
      status: InvoiceStatus.PAID,
      paidAt: new Date(),
      // Update paymentMethod jika kasir memilih metode bayar baru
      paymentMethod: input?.paymentMethod ?? invoice.paymentMethod,
    },
  });

  io.emit('QUEUE_UPDATED', { type: 'INVOICE_PAID', invoiceId: id });

  return payInvoice;
};

// CUSTOMERS
export const payCustomerInvoiceService = async (
  userId: number,
  invoiceId: number,
  paymentMethod: PaymentMethod = PaymentMethod.QRIS,
) => {
  // ambil data pasien milik user
  const patient = await prisma.patient.findUnique({
    where: { userId },
  });
  if (!patient) throw new ApiError(404, 'Data pasien tidak ditemukan');

  // ambil invoice dan pastikan invoice ini benar milik pasien yang sedang login (security check)
  const invoice = await prisma.invoice.findUnique({
    where: { id: invoiceId },
    include: { visit: true },
  });

  if (!invoice) throw new ApiError(404, 'Tagihan invoice tidak di temukan');

  if (invoice.visit.patientId !== patient.id)
    throw new ApiError(403, 'Anda tidak memiliki hak akses untuk membayar tagihan pasien lain');

  if (invoice.status === InvoiceStatus.PAID) throw new ApiError(400, 'Tagihan ini sudah lunas');

  // update status invoice menjadi PAID
  const updateInvoice = await prisma.invoice.update({
    where: { id: invoiceId },
    data: {
      status: InvoiceStatus.PAID,
      paymentMethod,
      paidAt: new Date(),
    },
  });

  io.emit('QUEUE_UPDATED', { type: 'INVOICE_PAID', invoiceId });

  return updateInvoice;
};

// MIDTRANS
// request snap token ke midtrans
export const getMidtransSnapTokenService = async (invoiceId: number, userId?: number) => {
  const invoice = await getInvoiceByIdService(invoiceId);

  if (invoice.status === InvoiceStatus.PAID) {
    throw new ApiError(400, 'Tagihan invoice ini sudah lunas sebelumnya');
  }

  // jika di panggil oleh customer, pastikan ini tagihan miliknya
  if (userId) {
    const patient = await prisma.patient.findUnique({
      where: { userId },
    });
    if (!patient || invoice.visit.patient.noRm !== patient.noRm) {
      throw new ApiError(403, 'Anda tidak memiliki hak akses tagihan ini');
    }
  }

  // rincian item invoice buat di kirim ke midtrans
  const itemDetails: Array<{ id: string; price: number; quantity: number; name: string }> = [
    {
      id: 'FEE-DOKTER',
      price: Math.round(Number(invoice.totalConsultationFee)),
      quantity: 1,
      name: `Jasa dokter (${invoice.visit.doctor.name})`.slice(0, 50),
    },
  ];

  if (invoice.visit.consultation?.consultationMedicines) {
    for (const m of invoice.visit.consultation.consultationMedicines) {
      itemDetails.push({
        id: `MED-${m.medicine.id || m.id}`,
        price: Math.round(Number(m.price)),
        quantity: m.qty,
        name: `${m.medicine.name}`.slice(0, 50),
      });
    }
  }

  const transactionDetails = {
    // tambahkan timestamp di order_id agar midtrans sandbox selalu ngizinin generate token baru jika pasien entry
    order_id: `${invoice.invoiceNo}-${Date.now()}`,
    gross_amount: Math.round(Number(invoice.totalAmount)),
  };

  const customerDetails = {
    first_name: invoice.visit.patient.name,
    phone: invoice.visit.patient.phone || '08123456789',
  };

  // minta snap token dari midtrans
  const parameter = {
    transaction_details: transactionDetails,
    item_details: itemDetails,
    customer_details: customerDetails,
  };

  const snapTransaction = await snap.createTransaction(parameter);

  return {
    token: snapTransaction.token,
    redirectUrl: snapTransaction.redirect_url,
    invoiceNo: invoice.invoiceNo,
    totalAmount: invoice.totalAmount,
  };
};

// webhook notifikasi handler (dipanggil oleh midtrans saat bayar lunas)
export const handleMidtransNotificationService = async (notificationPayload: any) => {
  console.log('>>> [MIDTRANS WEBHOOK RECEIVED]:', JSON.stringify(notificationPayload));

  let statusResponse = notificationPayload;
  try {
    statusResponse = await snap.transaction.notification(notificationPayload);
  } catch (err) {
    console.warn('snap.transaction.notification warning, using raw payload:', err);
  }

  const orderId = statusResponse.order_id || notificationPayload.order_id;
  const transactionStatus = statusResponse.transaction_status || notificationPayload.transaction_status;
  const fraudStatus = statusResponse.fraud_status || notificationPayload.fraud_status;

  console.log(`>>> [MIDTRANS STATUS]: orderId=${orderId}, status=${transactionStatus}, fraud=${fraudStatus}`);

  if (!orderId) {
    return { status: 'IGNORED_NO_ORDER_ID' };
  }

  // Extract InvoiceNo asli (buang timestamp di bagian paling belakang)
  // Contoh: "INV-2025-0021-1788229937614" -> "INV-2025-0021"
  const invoiceNo = orderId.includes('-')
    ? orderId.substring(0, orderId.lastIndexOf('-'))
    : orderId;

  const invoice =
    (await prisma.invoice.findFirst({
      where: { invoiceNo },
    })) ||
    (await prisma.invoice.findFirst({
      where: { invoiceNo: orderId },
    }));

  if (!invoice) {
    console.error(`Invoice ${invoiceNo} tidak ditemukan di database`);
    return { status: 'INVOICE_NOT_FOUND', invoiceNo };
  }

  // Cek apakah status transaksi sukses (settlement atau capture)
  if (transactionStatus === 'settlement' || (transactionStatus === 'capture' && fraudStatus === 'accept')) {
    if (invoice.status !== InvoiceStatus.PAID) {
      const paymentType = (statusResponse.payment_type || notificationPayload.payment_type || '').toLowerCase();
      const mappedPaymentMethod =
        paymentType.includes('qris') || paymentType.includes('gopay') || paymentType.includes('shopeepay')
          ? PaymentMethod.QRIS
          : paymentType.includes('card')
          ? PaymentMethod.CARD
          : PaymentMethod.TRANSFER;

      await prisma.invoice.update({
        where: { id: invoice.id },
        data: {
          status: InvoiceStatus.PAID,
          paymentMethod: mappedPaymentMethod,
          paidAt: new Date(),
        },
      });

      console.log(`>>> [INVOICE PAID SUCCESS]: Invoice ${invoice.invoiceNo} marked as PAID via ${mappedPaymentMethod}`);
      // Broadcast Socket.IO agar Dashboard & HP Pasien auto-update ke Farmasi!
      io.emit('QUEUE_UPDATED', { type: 'INVOICE_PAID', invoiceId: invoice.id });
    }
  }

  return { status: 'OK' };
};

File: clinic_backend_go/internal/service/invoice_service.go
package service

import (
	"bytes"
	"clinic_backend_go/internal/models"
	"clinic_backend_go/internal/repository"
	"clinic_backend_go/internal/ws"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"time"
)

type PayInvoiceInput struct {
	PaymentMethod    string `json:"paymentMethod"`
	PaymentMethodAlt string `json:"payment_method"`
}

func (p *PayInvoiceInput) GetPaymentMethod() string {
	if p.PaymentMethod != "" {
		return p.PaymentMethod
	}
	if p.PaymentMethodAlt != "" {
		return p.PaymentMethodAlt
	}
	return "CASH"
}

type CreateInvoiceInput struct {
	VisitID              uint    `json:"visitId"`
	VisitIDAlt           uint    `json:"visit_id"`
	TotalConsultationFee float64 `json:"total_consultation_fee"`
	TotalMedicineFee     float64 `json:"total_medicine_fee"`
	TotalAmount          float64 `json:"total_amount"`
	PaymentMethod        string  `json:"paymentMethod"`
	PaymentMethodAlt     string  `json:"payment_method"`
}

func (c *CreateInvoiceInput) GetVisitID() uint {
	if c.VisitID > 0 {
		return c.VisitID
	}
	return c.VisitIDAlt
}

type MidtransSnapResponse struct {
	Token       string  `json:"token"`
	RedirectURL string  `json:"redirectUrl"`
	InvoiceNo   string  `json:"invoiceNo"`
	TotalAmount float64 `json:"totalAmount"`
}

type MidtransItemDetail struct {
	ID       string `json:"id"`
	Price    int64  `json:"price"`
	Quantity int    `json:"quantity"`
	Name     string `json:"name"`
}

type MidtransTransactionDetails struct {
	OrderID     string `json:"order_id"`
	GrossAmount int64  `json:"gross_amount"`
}

type MidtransCustomerDetails struct {
	FirstName string `json:"first_name"`
	Phone     string `json:"phone"`
}

type MidtransSnapPayload struct {
	TransactionDetails MidtransTransactionDetails `json:"transaction_details"`
	ItemDetails        []MidtransItemDetail       `json:"item_details"`
	CustomerDetails    MidtransCustomerDetails    `json:"customer_details"`
}

type InvoiceService interface {
	GetAllInvoices() ([]models.Invoice, error)
	GetInvoiceByID(id uint) (*models.Invoice, error)
	CreateInvoice(input CreateInvoiceInput) (*models.Invoice, error)
	PayInvoice(id uint, method string) (*models.Invoice, error)
	GetMidtransSnapToken(invoiceID uint, userID uint, userRole string) (*MidtransSnapResponse, error)
}

type invoiceService struct {
	invoiceRepo repository.InvoiceRepository
	visitRepo   repository.VisitRepository
	patientRepo repository.PatientRepository
	wsHub       *ws.Hub
}

func NewInvoiceService(invoiceRepo repository.InvoiceRepository, visitRepo repository.VisitRepository, patientRepo repository.PatientRepository, wsHub *ws.Hub) InvoiceService {
	return &invoiceService{invoiceRepo: invoiceRepo, visitRepo: visitRepo, patientRepo: patientRepo, wsHub: wsHub}
}

func (s *invoiceService) GetAllInvoices() ([]models.Invoice, error) {
	return s.invoiceRepo.FindAll()
}

func (s *invoiceService) GetInvoiceByID(id uint) (*models.Invoice, error) {
	return s.invoiceRepo.FindByID(id)
}

func (s *invoiceService) CreateInvoice(input CreateInvoiceInput) (*models.Invoice, error) {
	visitID := input.GetVisitID()
	if visitID == 0 {
		return nil, errors.New("visitId atau visit_id wajib diisi")
	}

	// Jika invoice untuk kunjungan ini sudah dibuat (misal dari sesi konsultasi), kembalikan langsung
	existing, _ := s.invoiceRepo.FindByVisitID(visitID)
	if existing != nil {
		return existing, nil
	}

	visit, err := s.visitRepo.FindByID(visitID)
	if err != nil {
		return nil, errors.New("data kunjungan tidak ditemukan")
	}

	totalAmount := input.TotalAmount
	if totalAmount == 0 {
		totalAmount = input.TotalConsultationFee + input.TotalMedicineFee
	}

	method := input.PaymentMethod
	if method == "" {
		method = input.PaymentMethodAlt
	}
	if method == "" {
		method = "CASH"
	}

	now := time.Now()
	invoice := models.Invoice{
		VisitID:              visit.ID,
		InvoiceNo:            fmt.Sprintf("INV-%d-%d", now.Unix(), visit.ID),
		TotalConsultationFee: input.TotalConsultationFee,
		TotalMedicineFee:     input.TotalMedicineFee,
		TotalAmount:          totalAmount,
		Status:               "UNPAID",
		PaymentMethod:        method,
		CreatedAt:            now,
		UpdatedAt:            now,
	}

	if err := s.invoiceRepo.Create(&invoice); err != nil {
		return nil, err
	}
	return &invoice, nil
}

func (s *invoiceService) PayInvoice(id uint, method string) (*models.Invoice, error) {
	invoice, err := s.invoiceRepo.FindByID(id)
	if err != nil {
		return nil, errors.New("invoice tidak ditemukan")
	}

	now := time.Now()
	invoice.Status = "PAID"
	invoice.PaymentMethod = method
	invoice.PaidAt = &now
	invoice.UpdatedAt = now

	if err := s.invoiceRepo.Update(invoice); err != nil {
		return nil, err
	}

	// Update status kunjungan menjadi COMPLETED
	_ = s.visitRepo.UpdateStatus(invoice.VisitID, "COMPLETED")

	if s.wsHub != nil {
		s.wsHub.BroadcastQueue(map[string]interface{}{
			"type":      "INVOICE_PAID",
			"invoiceId": invoice.ID,
			"visitId":   invoice.VisitID,
		})
	}

	return invoice, nil
}

func (s *invoiceService) GetMidtransSnapToken(invoiceID uint, userID uint, userRole string) (*MidtransSnapResponse, error) {
	invoice, err := s.invoiceRepo.FindByID(invoiceID)
	if err != nil || invoice == nil {
		return nil, errors.New("Tagihan invoice tidak ditemukan")
	}

	if invoice.Status == "PAID" {
		return nil, errors.New("Tagihan invoice ini sudah lunas sebelumnya")
	}

	// Jika pemanggil adalah pasien (CUSTOMER), pastikan ini tagihan miliknya
	if userRole == "CUSTOMER" && userID > 0 {
		patient, err := s.patientRepo.FindByUserID(userID)
		if err != nil || patient == nil || (invoice.Visit != nil && invoice.Visit.PatientID != patient.ID) {
			return nil, errors.New("Anda tidak memiliki hak akses tagihan ini")
		}
	}

	var itemDetails []MidtransItemDetail
	doctorName := "Dokter"
	patientName := "Pasien"
	patientPhone := "08123456789"

	if invoice.Visit != nil {
		if invoice.Visit.Doctor != nil && invoice.Visit.Doctor.Name != "" {
			doctorName = invoice.Visit.Doctor.Name
		}
		if invoice.Visit.Patient != nil {
			if invoice.Visit.Patient.Name != "" {
				patientName = invoice.Visit.Patient.Name
			}
			if invoice.Visit.Patient.Phone != nil && *invoice.Visit.Patient.Phone != "" {
				patientPhone = *invoice.Visit.Patient.Phone
			}
		}
	}

	// 1. Fee Konsultasi Dokter
	consultationFee := int64(math.Round(invoice.TotalConsultationFee))
	if consultationFee > 0 {
		feeName := fmt.Sprintf("Jasa Dokter (%s)", doctorName)
		if len(feeName) > 50 {
			feeName = feeName[:50]
		}
		itemDetails = append(itemDetails, MidtransItemDetail{
			ID:       "FEE-DOKTER",
			Price:    consultationFee,
			Quantity: 1,
			Name:     feeName,
		})
	}

	// 2. Obat-obatan (jika ada)
	if invoice.Visit != nil && invoice.Visit.Consultation != nil {
		for _, m := range invoice.Visit.Consultation.ConsultationMedicines {
			medName := "Obat"
			if m.Medicine != nil && m.Medicine.Name != "" {
				medName = m.Medicine.Name
			}
			if len(medName) > 50 {
				medName = medName[:50]
			}
			itemDetails = append(itemDetails, MidtransItemDetail{
				ID:       fmt.Sprintf("MED-%d", m.MedicineID),
				Price:    int64(math.Round(m.Price)),
				Quantity: m.Qty,
				Name:     medName,
			})
		}
	}

	grossAmount := int64(math.Round(invoice.TotalAmount))

	// Midtrans mewajibkan gross_amount == sum(itemDetails.price * quantity)
	var sumItems int64
	for _, item := range itemDetails {
		sumItems += item.Price * int64(item.Quantity)
	}

	if sumItems != grossAmount || len(itemDetails) == 0 {
		itemDetails = []MidtransItemDetail{
			{
				ID:       fmt.Sprintf("INV-%d", invoice.ID),
				Price:    grossAmount,
				Quantity: 1,
				Name:     fmt.Sprintf("Tagihan %s", invoice.InvoiceNo),
			},
		}
	}

	// Tambahkan unix timestamp pada order_id agar selalu generate token baru di sandbox
	orderID := fmt.Sprintf("%s-%d", invoice.InvoiceNo, time.Now().Unix())

	payload := MidtransSnapPayload{
		TransactionDetails: MidtransTransactionDetails{
			OrderID:     orderID,
			GrossAmount: grossAmount,
		},
		ItemDetails: itemDetails,
		CustomerDetails: MidtransCustomerDetails{
			FirstName: patientName,
			Phone:     patientPhone,
		},
	}

	bodyBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	serverKey := os.Getenv("MIDTRANS_SERVER_KEY")
	if serverKey == "" {
		serverKey = "SB-Mid-server-wQg5dLb3EIDT4kWSGKh4qv6y"
	}
	isProd := os.Getenv("MIDTRANS_IS_PRODUCTION") == "true"
	snapURL := "https://app.sandbox.midtrans.com/snap/v1/transactions"
	if isProd {
		snapURL = "https://app.midtrans.com/snap/v1/transactions"
	}

	req, err := http.NewRequest("POST", snapURL, bytes.NewBuffer(bodyBytes))
	if err != nil {
		return nil, err
	}

	authHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(serverKey+":"))
	req.Header.Set("Authorization", authHeader)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("gagal menghubungi Midtrans: %w", err)
	}
	defer resp.Body.Close()

	respBytes, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("midtrans error (%d): %s", resp.StatusCode, string(respBytes))
	}

	var snapResp struct {
		Token       string `json:"token"`
		RedirectURL string `json:"redirect_url"`
	}
	if err := json.Unmarshal(respBytes, &snapResp); err != nil {
		return nil, fmt.Errorf("gagal membaca respon Midtrans: %w", err)
	}

	return &MidtransSnapResponse{
		Token:       snapResp.Token,
		RedirectURL: snapResp.RedirectURL,
		InvoiceNo:   invoice.InvoiceNo,
		TotalAmount: invoice.TotalAmount,
	}, nil
}
